File: CS/ViewModel.cs
using DevExpress.Mvvm;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace MasterDetailInside {
    public class ParentDataItem : BindableBase {
        public string Text { get { return GetValue<string>(); } set { SetValue(value); } }

        public int Number { get { return GetValue<int>(); } set { SetValue(value); } }

        public ObservableCollection<DataItem> Data { get { return GetValue<ObservableCollection<DataItem>>(); } set { SetValue(value); } }
    }

    public class DataItem : BindableBase {
        public bool Ready { get { return GetValue<bool>(); } set { SetValue(value); } }

        public string Text { get { return GetValue<string>(); } set { SetValue(value); } }

        public int Number { get { return GetValue<int>(); } set { SetValue(value); } }
    }

    public class ViewModel {
        public ObservableCollection<ParentDataItem> Data { get; set; }

        public ViewModel() {
            Data = new ObservableCollection<ParentDataItem>(GetData());
        }

        static IEnumerable<ParentDataItem> GetData() {
            for(int i = 0; i < 100; i++) {
                var parentTestData = new ParentDataItem() { Text = "Master" + i, Number = i, Data = new ObservableCollection<DataItem>() };
                for(int j = 0; j < 50; j++) {
                    parentTestData.Data.Add(new DataItem() { Text = "Detail" + j + " Master" + i, Number = j, Ready = j % 2 != 0 });
                }
                yield return parentTestData;
            }
        }
    }
}

File: VB/ViewModel.vb
Imports DevExpress.Mvvm
Imports System.Collections.ObjectModel

Namespace MasterDetailInside

    Public Class ViewModel

        Private dataField As ObservableCollection(Of ParentTestData)

        Public ReadOnly Property Data As ObservableCollection(Of ParentTestData)
            Get
                If dataField Is Nothing Then
                    dataField = New ObservableCollection(Of ParentTestData)()
                    For i As Integer = 0 To 100 - 1
                        Dim parentTestData As ParentTestData = New ParentTestData() With {.Text = "Master" & i, .Number = i, .Data = New ObservableCollection(Of TestData)()}
                        For j As Integer = 0 To 50 - 1
                            parentTestData.Data.Add(New TestData() With {.Text = "Detail" & j & " Master" & i, .Number = j, .Ready = j Mod 2 <> 0})
                        Next

                        dataField.Add(parentTestData)
                    Next
                End If

                Return dataField
            End Get
        End Property
    End Class

    Public Class TestData
        Inherits BindableBase

        Public Property Ready As Boolean
            Get
                Return GetProperty(Function() Me.Ready)
            End Get

            Set(ByVal value As Boolean)
                SetProperty(Function() Ready, value)
            End Set
        End Property

        Public Property Text As String
            Get
                Return GetProperty(Function() Me.Text)
            End Get

            Set(ByVal value As String)
                SetProperty(Function() Text, value)
            End Set
        End Property

        Public Property Number As Integer
            Get
                Return GetProperty(Function() Me.Number)
            End Get

            Set(ByVal value As Integer)
                SetProperty(Function() Number, value)
            End Set
        End Property
    End Class

    Public Class ParentTestData
        Inherits BindableBase

        Public Property Text As String
            Get
                Return GetProperty(Function() Me.Text)
            End Get

            Set(ByVal value As String)
                SetProperty(Function() Text, value)
            End Set
        End Property

        Public Property Number As Integer
            Get
                Return GetProperty(Function() Me.Number)
            End Get

            Set(ByVal value As Integer)
                SetProperty(Function() Number, value)
            End Set
        End Property

        Public Property Data As ObservableCollection(Of TestData)
            Get
                Return GetProperty(Function() Me.Data)
            End Get

            Set(ByVal value As ObservableCollection(Of TestData))
                SetProperty(Function() Data, value)
            End Set
        End Property
    End Class
End Namespace
